File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        Map<String, Integer> list = new HashMap<>();
        for(int i=0; i<want.length; i++){
            //키-값 쌍으로 저장
            list.put(want[i], number[i]);
        }
        
        
        int answer = 0;
        
        //슬라이딩 윈도우 기법
        for(int i=0; i<=discount.length-10; i++){
            Map<String, Integer> check = new HashMap<>();
            
            for(int j=i; j<i+10; j++){
                if(list.containsKey(discount[j])){
                    check.put(discount[j], check.getOrDefault(discount[j], 0)+1);
                }
            }
            
            if(check.equals(list)) answer++;
        }
        
        
        return answer;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
# 슬라이딩 윈도우
def solution(want, number, discount):
    answer = 0
    start = 0
    end = 10
    want_dict = {}
    for i in range(len(want)):
        want_dict[want[i]] = number[i]
    
    if len(discount) < 10:
        return 0
    
    while end <= len(discount):
        discount_dict = {}
        for i in range(start, end):
            # want의 개수와 일치하는지 확인
            discount_dict[discount[i]] = discount_dict.get(discount[i], 0) + 1
        if want_dict == discount_dict:
            answer += 1
            
        start += 1
        end += 1
        
    return answer